### DISCUSSION TITLE
Discriminators as first class citizens of sum types

### DISCUSSION THREAD

**Discussion by goshacodes** [scala/scala3#21205] [Feature Requests]
Consider following AST:
```scala 3
enum Foo(val discriminator: String):
  case Bar() extends Foo("bar")
  case Baz() extends Foo("baz")
```

"bar" and "baz" are literal constants, but now they are not exposed at compile time in any way.

So to make derivation of JsonReader (for example) possible I need to stub all case children instances and get this field.

I propose to:
1. Make this literal constants (if they are literal constants for all of the case children) visible at compile time in any way
2. If possible - expose this information to mirrors

So generation of this code could be possible without dirty hacks
```scala
given barJsonReader: JsonReader[Bar] = ???
given bazJsonReader: JsonReader[Baz] = ???

val discriminator = ???

discriminator match
  case "bar" =>
    barJsonReader
  case "baz" =>
    bazJsonReader
  
```

Why not use ordinal or MirroredElemLabels:

1. Enum ordinal is bound to ordinal, consider situation when developer changes order of case chilren (just because he doesn't like it) and deserialization starts failing at runtime

2. When you write integration with other software systems you often use provided by that system protocol, where discriminators can be strings and not bound in any way to ordinal of case children in your scala model

3. Case labels have same problem, they are bound to your case class name

**Comment by bishabosha** [comment]
> 2. If possible - expose this information to mirrors

The ordinal of an enum case is compile time information - its the index of the type in the MirroredElemTypes in the Mirror - which you can compute manually, or its the same as `Tuple.Index[MirroredElemTypes, Baz]`

**Reply by goshacodes** [reply]
I've changed description a bit, discriminator has different type. Ordinal of enum cases doesn't always match discriminator value

**Reply by bishabosha** [reply]
ok, you haven't really explained why you need your custom value at compiletime to "discriminate" when ordinal already exists as a discriminator

**Reply by bishabosha** [reply]
I think you are trying to have some "$type" field in a json object? the case labels of the types are also compile time strings accessible through the mirror - so you could use those

**Reply by goshacodes** [reply]
1. Enum ordinal is bound to ordinal, consider situation when developer changes order of case chilren (just because he doesn't like it) and deserialization starts failing at runtime

2. When you write integration with other software systems you often use provided by that system protocol, where discriminators can be strings and not bound in any way to ordinal of case children in your scala model

**Reply by goshacodes** [reply]
Case labels have same problem, they are bound to your case class name

**Reply by goshacodes** [reply]
Yes
> I think you are trying to have some "$type" field in a json object?

**Reply by bishabosha** [reply]
you can use an annotation? those are compile time information

**Reply by goshacodes** [reply]
I probably can, but this is not what I want. Situation I mention is very common and annotations are clumsy

**Reply by goshacodes** [reply]
I think this could be even thought as a language construct. Because this can really simplify life for both users and maintainers.
E.g. new keyword `selector`.
```scala 3
sealed trait Foo(selector name: String)
```
It is bound to sealed trait and enums parameter.
Only one field can be a selector
If it is used - all cases should provide this field as literal constant during `extends`
Information about this field can be accessed compile time

**Reply by bishabosha** [reply]
If we dont do that, then we already have a new `tracked` modifier for parameters under `language.experimental.modularity` - which can refine singleton types such as literal strings. combined with some parent class like `DiscriminatedEnum` it could define a parameter that a mirror could look for, and perform the uniqueness check. - I tried this in https://github.com/scala/scala3/issues/21213 (with a bug)

**Reply by bishabosha** [reply]
or we could tell the mirror to specifically look for a `tracked` parameter of a specific name - which would have less boilerplate

**Reply by bishabosha** [reply]
but I do like the idea of the selector modifier, maybe worth a sip proposal

**Reply by goshacodes** [reply]
Let's try SIP

**Reply by bishabosha** [reply]
Just coming back to this - why would the enum case name not be appropriate? just because its inconvenient to be lower case? ( i.e. why would the selector be different than the case name?)

**Reply by goshacodes** [reply]
Yes, you may end up making it snake_case or kebab-case with backticks just because in other contexts someone use it like this. And backticks are not very convenient to use, but thats not all.

I was grown in a paradigm where you shouldn't use a class name, because this may break compatibility when any class name change and developers may change names just because they want to.

Or it maybe not just basic type, the discriminator can be enum itself or something else with provided typeclass instance for it.

Would be great to have a way to get those values compile time in macro. Currently it is not possible - you need an instance to get those.

```scala
enum Foo(val discriminator: String):
  case Bar() extends Foo("bar")
  case Baz() extends Foo("baz")
```

Maybe it is not a big problem, but one day I thought it could be useful.

**Reply by prolativ** [reply]
How about using singleton type parameters as a workaround in this case? E.g.
```scala
enum Foo[D <: String & Singleton]:
  case Bar() extends Foo["bar"]
  case Baz() extends Foo["baz"]

  inline def discriminator: D = compiletime.constValue[D]

object Foo:
  inline def caseDiscriminator[F <: Foo[?]]: String =
    inline compiletime.erasedValue[F] match
      case _: Foo[d] => compiletime.constValue[d]

@main def run() =
  println(Foo.Bar().discriminator)
  println(Foo.caseDiscriminator[Foo.Bar])
```
Unfortunately I didn't manage to make this work with type members, which would save us from writing things like `Foo[?]`

**Reply by goshacodes** [reply]
It brings complexity to users, I wouldn't use it myself
